Framework/CMakeLists.txt

```diff
@@ -1,232 +1,229 @@
+
+ROOT_GENERATE_DICTIONARY("${dict}" ${HEADERS}
+  LINKDEF include/QualityControl/LinkDef.h)
+
+# TODO review how and what to install for dictionary
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/${LIBRARY_NAME_CORE}Dict_rdict.pcm
+  ${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/${LIBRARY_NAME_CORE}Dict.rootmap
+  DESTINATION lib)
 
-O2_GENERATE_LIBRARY()
+####################################
+# Library
+####################################
 
-# Task Launcher
-O2_GENERATE_EXECUTABLE(
-        EXE_NAME qcTaskLauncher
-        SOURCES src/TaskDevice.cxx
-        MODULE_LIBRARY_NAME ${LIBRARY_NAME}
-        BUCKET_NAME ${BUCKET_NAME}
-)
+# Create library
+add_library(QualityControl SHARED ${SRCS} QualityControlDict.cxx)
 
-O2_GENERATE_EXECUTABLE(
-        EXE_NAME qcCheckerLauncher
-        SOURCES src/qcCheckerLauncher.cxx
-        MODULE_LIBRARY_NAME ${LIBRARY_NAME}
-        BUCKET_NAME ${BUCKET_NAME}
+target_include_directories(QualityControl
+  PUBLIC
+    $<INSTALL_INTERFACE:include>
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+  PRIVATE
+    ${CMAKE_CURRENT_SOURCE_DIR}/src
 )
 
-O2_GENERATE_EXECUTABLE(
-        EXE_NAME qcSpy
-        SOURCES src/qcSpy.cxx
-        MODULE_LIBRARY_NAME ${LIBRARY_NAME}
-        BUCKET_NAME ${BUCKET_NAME}
+# Link targets
+target_link_libraries(QualityControl
+  PUBLIC
+    Boost::boost
+    FairLogger::FairLogger
+    DataSampling::DataSampling
+    FairMQ::FairMQ
+    ROOT::Hist
+    AliceO2::Common
+    AliceO2::InfoLogger
+    Monitoring::Monitoring
+    Configuration::Configuration
+    ROOT::Net
+    AliceO2::AliceO2
+    Boost::container
+  PRIVATE
+    Boost::system
+    MySQL::MySQL
+    ROOT::Gui
+    ROOT::RMySQL
+    CURL::CURL
 )
 
-O2_GENERATE_EXECUTABLE(
-        EXE_NAME qcInfoService
-        SOURCES src/runInformationService.cxx
-        MODULE_LIBRARY_NAME ${LIBRARY_NAME}
-        BUCKET_NAME ${BUCKET_NAME}
-)
+####################################
+# Executables
+####################################
 
-O2_GENERATE_EXECUTABLE(
-        EXE_NAME qcInfoServiceDump
-        SOURCES src/runInformationServiceDump.cxx
-        MODULE_LIBRARY_NAME ${LIBRARY_NAME}
-        BUCKET_NAME ${BUCKET_NAME}
+set(EXE_SRCS
+  src/TaskDevice.cxx
+  src/qcCheckerLauncher.cxx
+  src/qcSpy.cxx
+  src/runInformationService.cxx
+  src/runInformationServiceDump.cxx
+  src/runTaskDPL.cxx
+  src/runReadoutChainTemplate.cxx
+  src/runMergerTest.cxx
+  src/runReadoutDataSampling.cxx
+  src/alfaTestReceiver.cxx
+  src/Consumer.cxx
+  src/runRepositoryBenchmark.cxx
 )
 
-O2_GENERATE_EXECUTABLE(
-        EXE_NAME taskDPL
-        SOURCES src/runTaskDPL.cxx
-        MODULE_LIBRARY_NAME ${LIBRARY_NAME}
-        BUCKET_NAME ${BUCKET_NAME}
+set(EXE_NAMES
+  qcTaskLauncher
+  qcCheckerLauncher
+  qcSpy
+  qcInfoService
+  qcInfoServiceDump
+  taskDPL
+  runReadoutChainTemplate
+  runMergerTest
+  runReadoutDataSampling
+  alfaTestReceiver
+  qcConsumer
+  repositoryBenchmark
```

Comment:
Just to understand what is going on here - the order of EXE_SRCS and EXE_NAMES has to match, right? Isn't there some dictionary-like way to do it?

Reply:
unfortunately I could not find a dictionary or map in CMake

Reply:
Still, the boilerplate reduction is very impressive!

Reply:
They don't exist. CMake does not even have lists natively... it uses `;`-separated strings...